### PR TITLE
Bootstrap style guide, best practices, philosophy, etc

### PR DESCRIPTION
**Pull request by hartz89**
This PR bootstraps the project.  Initial suite of markdown files covers things like:
- General philosophy
- "Code Gardening" manifesto
- Style guides: TypeScript, React, Vanilla Extract
- Best Practices: Testing, Refactoring
- Inspirations / attributions

### CODE REVIEW

File: best-practices/testing.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+- Prefer MSW (or equivalent) for HTTP — intercept at the network layer, not the `fetch` wrapper.
+- Don't mock internal modules. If the only way to test something is to mock an internal module, the test is probably at the wrong level.
+- "Reasonable" is a real constraint. Some SDKs don't expose a clean network seam; stop at the last layer that maps to something real.
+
+## Avoid Nesting
+
+`strength: strong` · [rationale](./testing.why.md#avoid-nesting)
+
+- Prefer flat `test` blocks.
+- One level of `describe` is usually enough.
```
Comment (hartz89) [inline best-practices/testing.md]: We should default to _no_ describe nesting. Reserve for cases like:
- A single module exporting several functions 
- A really complex integration suite with lots of behaviors to cover

But also keep in mind that both of the above are potential code smells.

File: best-practices/testing.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+test('advances from open to seed', () => {
+  const flower = makeFlower({ phase: 'open' });
+  advance(flower);
+  expect(flower.phase).toBe('seed');
+});
+```
+
+## Prefer Setup Functions
+
+`strength: strong` · [rationale](./testing.why.md#prefer-setup-functions)
```
Comment (hartz89) [inline best-practices/testing.md]: I'm afraid this will lead to "setup functions by default".  They're not always necessary, but often become necessary at scale.

File: best-practices/testing.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+- Prefer duplicated setup inline over shared `beforeEach` hooks.
+- When setup gets too repetitive to inline, extract a factory function: `makeFlower({ phase: 'bud' })`. Each test calls it explicitly.
+- Reserve `beforeEach` for fixture isolation (`jest.clearAllMocks()`, fake-timer resets), not for building domain objects.
+
+## Test Structure
+
+`strength: strong` · [rationale](./testing.why.md#test-structure)
+
+- Arrange, act, assert: set up, invoke, check. Separate sections with blank lines, not comments.
```
Comment (hartz89) [inline best-practices/testing.md:85]: This should be balanced with colocation in the context of fewer, longer tests: sometimes a test needs multiple iterations of arrange, act, assert.

File: best-practices/testing.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+- Prefer fewer, longer tests that walk a realistic flow over many tiny tests that each verify one thing.
+- A long integration test is still one test if it exercises one user story — even with multiple act-assert beats.
```
Comment (hartz89) [inline best-practices/testing.md]: I see that my previous comment is addressed here... maybe re-ordering / consolidating a bit would help?

File: best-practices/testing.why.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+The classic pyramid (many units, fewer integrations, few e2e) assumed integration tests meant spinning up a database and a server. With MSW, Testing Library, and modern runners, integration tests are fast enough that the pyramid flattens into a trophy: static base, integration body, thin e2e tip.
+
+### TypeScript is your base layer
+
+Types rule out entire classes of bugs before a test runs — wrong shapes, missing fields, impossible states. Writing unit tests for what the compiler already proves is duplicated effort that rots.
+
+### When to reach for each
+
+- **Static (TS)**: types, `unknown`, discriminated unions. Always.
```
Comment (hartz89) [inline best-practices/testing.why.md]: Please clarify this bullet.

File: best-practices/testing.why.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+- **Unit**: pure functions, reducers, parsers, algorithms. Any time logic is isolable from I/O.
```
Comment (hartz89) [inline best-practices/testing.why.md]: I primarily write unit tests for common utilities:
- as documentation
- to cover the edge cases that aren't reasonable to cover in integration tests
- to help quickly isolate a point of failure when a bunch of integration tests fail

We should not over-write unit tests; prefer integration.

File: best-practices/testing.why.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+**Strength:** strong
+
+### `beforeEach` hides dependencies
+
+A `beforeEach` block runs before every test in a `describe`, but you can't see it from the test itself. To understand what a test depends on, you have to scroll up — often across multiple levels of nesting — and trace every hook that applies. Factory functions make the dependency explicit: read the test top-down and see exactly what it sets up.
+
+### Factories are composable
+
+`makeFlower({ phase: 'open' })` accepts arguments. You can build variants inline, test-by-test, without a cascade of `describe` blocks each specializing setup. Composition scales; hook-layering doesn't.
```
Comment (hartz89) [inline best-practices/testing.why.md]: nit: The markdown frames these as "factories" but i usually think of them as "setup functions".  Factories _sometimes_ applies, other times it doesn't. Consider rephrasing.

File: code-style/_index.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
+# Code Style
+
+Opinionated guides for TypeScript, React, and Vanilla Extract CSS. Each guide is designed for fast scanning by humans and context-efficient consumption by LLMs.
```
Comment (hartz89) [inline code-style/_index.md]: Is the note about "optimized for LLMs" redundant with the root markdown(s)?

File: code-style/_index.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+Some guides split into two files: a rules file and a rationale file.
+
+- **`<lang>.md`** — rules and examples. Scannable format: strength tags (`non-negotiable` / `strong` / `weak`), concise rules, code samples. Rules link to rationale when the reasoning isn't obvious.
+- **`<lang>.why.md`** — expanded rationale and context. Section names match the rules file so anchor links work. Read this when you need to understand the _why_, defend an edge case, or evaluate a rule for your context.
+
+### Why split?
+
+**Scanning:** An LLM applying a guide only needs the rules and examples. Rationale prose (history, tradeoffs, exceptions) adds context cost without changing behavior — move it to a sibling file so the main file stays terse.
+
+**Reasoning:** A human (or an LLM making a judgment call) can read the why file to understand the full picture. Keeping rationale separate from rules means we can be opinionated in the rules without needing to hedge or explain every corner case inline.
```
Comment (hartz89) [inline code-style/_index.md]: Again: this is a general rule for the entire repo, not just the style guides. Should the explanation live here?  Apply same critical thought throughout, won't add further comments.

File: code-style/_index.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+## Adding a New Guide or Rule
+
+1. Create a new rules file (`<lang>.md`) or update an existing one.
+2. Tag each section with a strength level (see above).
+3. Add rationale only when it's non-obvious or worth defending. Keep rules terse; prose goes in the why file.
+4. If adding rationale, create a sibling `<lang>.why.md` file with section names that match the rules file exactly.
+5. In the rules file, add anchor links: `[rationale](./typescript.why.md#section-anchor)` so readers can jump to context.
+
+Naming convention: Section anchors use lowercase and hyphens (e.g., `#discriminated-unions`).
```
Comment (hartz89) [inline code-style/_index.md]: Should this stuff be moved to a contributing.md file, claude.md, or both?

File: code-style/react.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+export const Petal: FC<PetalProps> = memo(({ count }) => <span>{count} petals</span>);
+Petal.displayName = 'Petal';
+```
+
+## Event Handlers
+
+`strength: strong`
+
+- Prefix with `handle` (e.g., `handleClick`, `handleSubmit`).
```
Comment (hartz89) [inline code-style/react.md]: I prefer `on__`, e.g. `onClick`.  I'm willing to reconsider if there's a really strong argument to the contrary.  If updated, ensure that all examples using `handle__` are updated for consistency.

File: code-style/react.why.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+## Memoization
+
+**Rule:** With the React Compiler, skip manual memoization. Without it, use `memo` / `useCallback` / `useMemo` sparingly — only when there's a measurable reason.
+
+**Strength:** strong
+
+### Manual memoization is easy to get wrong
+
+`memo`, `useCallback`, and `useMemo` all take dependency arrays and referential-equality rules that are easy to miscount. A stale dep list silently skips updates; a missed dep causes infinite re-renders. The cost of getting it wrong is real, and most of the time the "optimization" saves nothing.
```
Comment (hartz89) [inline code-style/react.why.md]: How important is this callout?  I assume most projects will have React eslint plugins to ensure exhaustive deps.  Perhaps call out linting here?

File: code-style/typescript.md [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+};
+```
+
+## Avoid `any`; Prefer `unknown`
+
+`strength: non-negotiable` · [rationale](./typescript.why.md#any-vs-unknown)
+
+- Never use `any` without a comment explaining why.
+- Use `unknown` for values of truly unknown shape; narrow before use.
+- Escape hatches (`any`, `!`, `as X`) require a justifying comment.
```
Comment (hartz89) [inline code-style/typescript.md:140]: We should add stronger wording around avoiding `as` casting, with accompanying rationale (if none exists yet).  It undermines the type system, and the type system is only as strong as its weakest link.

File: code-style/typescript.md [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+// avoid — the layers hide the actual shape
+type WeirdFlower = Omit<Partial<Pick<Flower, keyof Rose>>, 'thorns'>;
+```
+
+## Generics
+
+`strength: strong`
+
+- Use descriptive names in public APIs; single letters OK in tight local helpers.
+- Constrain generics — an unconstrained `T` is usually too loose.
```
Comment (hartz89) [inline code-style/typescript.md]: Prefix generic names with `T`, e.g. `TItem`, `TKey`.  `T` alone is okay for simple cases.

File: code-style/typescript.md [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+```ts
+const find = (id: string): Flower | undefined => flowers.find((f) => f.id === id);
+const name = flower?.name ?? 'unknown';
+```
+
+## Function Signatures
+
+`strength: strong`
+
+- Prefer arrow functions at module scope.
```
Comment (hartz89) [inline code-style/typescript.md]: `function` declarations are great for hoisting (e.g. putting a locally-scope utility function towards the bottom of a file for readability; only parse it when you need to kind of thing).  Is that a bad opinion?

**Review by hartz89 — COMMENTED** [review]
- Typescript: I think we're missing callouts about function parameters. Arguments lists are certainly not a bad thing in all cases, but for complex functions, object parameters are useful.  Often I'll have a "primary argument" (e.g. the thing we're mapping from in a mapping fn) and an "options" or "context" argument next.  Over-using argument lists can make the consumer side of code hard to read.  If my phrasing here isn't clear, pull external resources to help refine.
- Typescript: keep code as flat as reasonably possible. extracting complex conditional logic into helper functions can help maintain immutability and flatness. don't be afraid to reach for IIFEs for complex cases.
- Typescript: Strive for strict typescript. NPEs are just too easy without it. Sometimes less strictness in test files / mocks is warranted.
- React: Keep your JSX colocated in the return statement; don't declare JSX in multiple places.  Inline conditionals when possible.  Don't nest JSX deeply in helper functions / hooks away from the component body.  Sprinkling JSX outside the component return statement is a code smell which might indicate the need to break into sub-components.
- CSS (general - maybe we need a new guide?): if you plan to internationalize your app/package, don't forget about right-to-left and bi-directional languages. Prefer logical directions over physical directions in CSS.
- if we do introduce a new general guide for CSS, evaluate the vanilla extract CSS file for possible extractions.
